### PR TITLE
Welcome screen shows "Continue with authorization" when OSM account created from authorization flow

### PR DESCRIPTION
**Pull request by milan-cvetkovic**
When user wants to authorize 3rd party app with OSM, and wants to create an OSM account during OAuth authorization, the welcome screen displays "Continue with authorization" instead of "Start mapping".

continue-auth

### CODE REVIEW

**Comment by mmd-osm** [comment]
I'm not exactly clear if this is meant as an alternative approach to the email notification discussed in #4246?

**Comment by milan-cvetkovic** [comment]
> I'm not exactly clear if this is meant as an alternative approach to the email notification discussed in #4246?

This PR is an alternative PR to https://github.com/openstreetmap/openstreetmap-website/pull/4288, both of them discussed in #4246.

**Comment by AntonKhorev** [comment]
Not always going to work, see https://github.com/openstreetmap/openstreetmap-website/issues/4246#issuecomment-1749150883.

**Comment by milan-cvetkovic** [comment]
> Not always going to work, see [#4246 (comment)](https://github.com/openstreetmap/openstreetmap-website/issues/4246#issuecomment-1749150883).

I tested this to work, please see https://github.com/openstreetmap/openstreetmap-website/issues/4246#issuecomment-1801412284

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -317,6 +305,24 @@ def auth_failure
 
   private
 
+  def welcome_options
+    uri = URI(session[:referer]) if session[:referer].present?
+    welcome_options = {}
+    welcome_options["oauth_return_url"] = session[:referer] if uri&.path == oauth_authorization_path
```
[the remark this answers is not in the document]
Comment (milan-cvetkovic) [reply app/controllers/users_controller.rb]: I missed this. Updated.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -317,6 +305,24 @@ def auth_failure
...
+
+    begin
+      %r{map=(.*)/(.*)/(.*)}.match(uri.fragment) do |m|
+        editor = Rack::Utils.parse_query(uri.query).slice("editor")
+        welcome_options = { "zoom" => m[1],
+                            "lat" => m[2],
+                            "lon" => m[3] }.merge(editor).merge(welcome_options)
```
[the remark this answers is not in the document]
Comment (milan-cvetkovic) [reply app/controllers/users_controller.rb:318]: I am not too sure about this.

What is the use case that would have a fragment in `session[:referrer]`?

I wasn't able to reproduce it...

[the remark this answers is not in the document]
Comment (milan-cvetkovic) [reply app/controllers/users_controller.rb]: @tomhughes I have separated the path to setting options for OAuth return URL and coordinates, as you suggested.

**Comment by milan-cvetkovic** [comment]
@tomhughes Is there anything else required for this PR to go through?

**Comment by tomhughes** [comment]
No I think that all looks good now, thanks.

**Pull request merged**